app/forms/declaration/GoodsLocation.scala

```diff
@@ -17,58 +17,76 @@
+    */
+  private val validateLocationType: String => Boolean = (input: String) => {
+    val correctLocationType: Set[String] = Set("A", "B", "C", "D")
+    val predicate = isContainedIn(correctLocationType)
+    input.drop(2).headOption.map(_.toString).exists(predicate)
+  }
+
+  /**
+    * Qualifier Code is defined in fourth characted in Location Code
+    */
+  private val validateQualifierCode: String => Boolean = (input: String) => {
+    val correctQualifierCode: Set[String] = Set("U", "Y")
+    val predicate = isContainedIn(correctQualifierCode)
+    input.drop(3).headOption.map(_.toString).exists(predicate)
+  }
 
   val mapping = Forms.mapping(
-    "typeOfLocation" -> text()
-      .verifying("supplementary.goodsLocation.typeOfLocation.empty", nonEmpty)
-      .verifying("supplementary.goodsLocation.typeOfLocation.error", isEmpty or (isAlphabetic and hasSpecificLength(1))),
-    "qualifierOfIdentification" -> text()
-      .verifying("supplementary.goodsLocation.qualifierOfIdentification.empty", nonEmpty)
-      .verifying("supplementary.goodsLocation.qualifierOfIdentification.error", isEmpty or (isAlphabetic and hasSpecificLength(1))),
-    "identificationOfLocation" -> optional(
-      text()
-        .verifying("supplementary.goodsLocation.identificationOfLocation.error", isEmpty or (isAlphanumeric and noLongerThan(35)))
-    ),
-    "additionalIdentifier" -> optional(
-      text()
-        .verifying("supplementary.goodsLocation.additionalIdentifier.error", isNumeric and noLongerThan(3))
-    ),
-    "addressLine" -> optional(
-      text().verifying("supplementary.goodsLocation.addressLine.error", isAlphanumericWithAllowedSpecialCharacters and noLongerThan(70))
-    ),
-    "city" -> optional(
-      text()
-        .verifying("supplementary.goodsLocation.city.error", isAlphanumericWithAllowedSpecialCharacters and noLongerThan(35))
-    ),
-    "postCode" -> optional(
-      text().verifying("supplementary.goodsLocation.postCode.error", isAlphanumericWithAllowedSpecialCharacters and noLongerThan(9))
-    ),
-    "country" ->
-      text()
-        .verifying("supplementary.address.country.empty", _.trim.nonEmpty)
-        .verifying("supplementary.address.country.error", input => input.trim.isEmpty || allCountries.exists(_.countryName == input))
-  )(GoodsLocation.apply)(GoodsLocation.unapply)
-
-  def form(): Form[GoodsLocation] = Form(mapping)
+    "code" -> text()
+      .verifying("declaration.goodsLocation.code.empty", nonEmpty)
+      .verifying(
+        "declaration.goodsLocation.code.error",
+        isEmpty or (
+          validateCountry and validateLocationType and validateQualifierCode and noShorterThan(10) and noLongerThan(17)
+        )
+      )
+  )(GoodsLocationForm.apply)(GoodsLocationForm.unapply)
+
+  def form(): Form[GoodsLocationForm] = Form(mapping)
+}
+
+case class GoodsLocation(country: String, typeOfLocation: String, qualifierOfIdentification: String, identificationOfLocation: String) {
```

Comment:
Maybe this class should leave in declaration exchange

Reply:
Good point. I'll move it there

Reply:
Moved